GEODE-4990: Cluster Config StartUp Race Condition

jujoramos opened the pull request:
GEODE-4990: Cluster Config StartUp Race Condition

- Added unit test for `GetClusterConfigurationFunction`.
- Added integration tests for `ClusterConfigurationLoader`.
- Set 'InternalLocator.isSharedConfigurationEnabled()' visibility
  as `public`.
- Set `ClusterConfigurationLoader.requestConfigurationFromOneLocator()`
  visibility as `protected`.
- `GetClusterConfigurationFunction` now throws an exception if the
  cluster configuraiton service is not enabled on the locator.
- `GetClusterConfigurationFunction` now returns null if the cluster
  configuraiton service is enabled but not yet started. The caller can
  decide whether to retry or fail fast.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

jinmeiliao commented on geode-core/src/main/java/org/apache/geode/management/internal/configuration/functions/GetClusterConfigurationFunction.java:
```diff
@@ -32,20 +30,35 @@
 
-    logger.info("Received request for configuration  : {}", groups);
+    // Return exception to the caller so startup fails fast.
+    if (!internalLocator.isSharedConfigurationEnabled()) {
```
should we check for isSharedConfigurationRunning() here also? If it's enabled but not in a running state yet, we should return this error too, right?

jujoramos replied:
Hello @jinmeiliao,

I don't think so, but I might be wrong here... the fact that the function is executed on a locator where the cluster configuration service is disabled (`internalLocator.isSharedConfigurationEnabled() == false`) can be considered as unexpected and that's the reason why the `IllegalStateException` is returned to the caller (`ClusterConfigurationLoader`).
It's possible (and expected if members are started up concurrently), however, for the cluster configuration service to be enabled but not yet running because it's actually starting up (`internalLocator.isSharedConfigurationEnabled() == true && internalLocator.isSharedConfigurationRunning() == false), in which case we shouldn't throw an exception but return `null` from the function. This allows the caller (`ClusterConfigurationLoader`) to decide whether to fail fast or keep retrying for some time, as it's currently doing [here](https://github.com/apache/geode/blob/d79282134c2064a09bfe418afcedaaf4c3741347/geode-core/src/main/java/org/apache/geode/internal/cache/ClusterConfigurationLoader.java#L278).
Please let me know your thoughts.
Cheers.

jujoramos replied:
Hey @jinmeiliao,

Thanks for the heads up, I haven't realized that we were already requesting the cluster configuration service *only* from locators where the service is actually enabled.
Either way, I have to disagree again... returning `null` when the cluster configuration service is not enabled can slow down the startup process altogether: the server will retry six times with a delay of 10 seconds after each retry, meaning that the startup failure will happen only after one minute. This shouldn't happen as we only request the cluster configuration from locators that have it enabled, but it's better to be on the safer side and fail fast on startup:

```
    int attempts = 6;
    OUTER: while (attempts > 0) {
      for (InternalDistributedMember locator : locatorList) {
        logger.info("Attempting to retrieve cluster configuration from {} - {} attempts remaining",
            locator.getName(), attempts);
        response = requestConfigurationFromOneLocator(locator, groups);
        if (response != null) {
          break OUTER;
        }
      }

      try {
        Thread.sleep(10000);
      } catch (InterruptedException ex) {
        break;
      }

      attempts--;
    }
```

Moreover, and thinking more about the `GetClusterConfigurationFunction` API, is more clear to return different things for different situations: if we return `null` on both scenarios (cluster configuration not enabled or cluster configuration enabled but not fully loaded yet) the user of the API (there's none at the moment other than `ClusterConfigurationLoader`, but it could be in the future...) doesn't get a clear perspective of what's happening... throwing an `IllegalStateException` when the cluster configuration service is not enabled and returning `null` when the service is enabled but not yet fully loaded is far more clear.
We could even refactor the `GetClusterConfigurationFunction` as a whole to return a typed response containing more accurate information about the status of the cluster configuration service without any exceptions at all, but that's another improvement, for the future maybe.
Please let me know what you think, I can make the changes if you still believe they are needed after reading the above.

PurelyApplied commented on geode-core/src/main/java/org/apache/geode/management/internal/configuration/functions/GetClusterConfigurationFunction.java:
```diff
@@ -32,20 +30,35 @@
+    if (!internalLocator.isSharedConfigurationEnabled()) {
+      String errorMessage = "The cluster configuration service is not enabled on this member.";
+      logger.warn(errorMessage);
+      context.getResultSender().lastResult(new IllegalStateException(errorMessage));
```
A matter of personal preference, but I think this would read better if this block explicitly `return`ed at the end instead of having a long `else` block following it.  The `try` block ends up being five tabs in and that just feels a little heavily nested for my taste.

jujoramos replied:
I hear you :-), the change is done.

PurelyApplied commented on geode-core/src/test/java/org/apache/geode/management/internal/configuration/functions/GetClusterConfigurationFunctionTest.java:
```diff
@@ -0,0 +1,128 @@
+import org.apache.geode.test.junit.categories.UnitTest;
+
+@Category(UnitTest.class)
+@PowerMockIgnore("*.UnitTest")
```
I'm not really familiar with `PowerMock`, but we don't have any classes that end with `.UnitTest`.  Or is that ignoring third-party classes?

PurelyApplied replied:
Huh.  Good to know!

PurelyApplied commented on geode-core/src/test/java/org/apache/geode/internal/cache/ClusterConfigurationLoaderIntegrationTest.java:
```diff
@@ -35,25 +41,100 @@
-  public void requestForClusterConfiguration() throws Exception {
+  public void requestConfigurationFromLocatorsShouldCorrectlyLoadTheClusterConfigurationEvenAfterSeveralRetries()
+      throws Exception {
+    int mockLimit = new Random().nextInt(6) + 1;
```
I really don't like seeing non-deterministic behavior in a test, beyond that which is inherent to the system.  Is there a reason this can't just be explicitly chosen?

jujoramos replied:
No reason at all, I've changed it to use the maximum amount of attempts.

PurelyApplied commented:
(I meant to make this a parent-comment to the review, but mis-clicked...)

The only thing I'm really concerned about is using `Random` in our tests.  The other two are meant to be comments more than _Requesting Changes_ items.

jujoramos commented on geode-core/src/main/java/org/apache/geode/internal/cache/ClusterConfigurationLoader.java:
```diff
@@ -308,7 +308,7 @@ public ConfigurationResponse requestConfigurationFromLocators(String groupList,
   }
 
-  private ConfigurationResponse requestConfigurationFromOneLocator(
+  protected ConfigurationResponse requestConfigurationFromOneLocator(
```
Hello @jinmeiliao,

Sorry to disagree again, but this is not accurate...
The code `Object result = ((ArrayList) resultCollector.getResult()).get(0);` ([line 318](https://github.com/jujoramos/geode/blob/719f261bfbe996f11386b33632d6debd3114a946/geode-core/src/main/java/org/apache/geode/internal/cache/ClusterConfigurationLoader.java#L318)) will actually throw a `FunctionException` and, thus, [line 324](https://github.com/jujoramos/geode/blob/719f261bfbe996f11386b33632d6debd3114a946/geode-core/src/main/java/org/apache/geode/internal/cache/ClusterConfigurationLoader.java#L324) doesn't come into play at all, the next execution step will be [line 329](https://github.com/jujoramos/geode/blob/719f261bfbe996f11386b33632d6debd3114a946/geode-core/src/main/java/org/apache/geode/internal/cache/ClusterConfigurationLoader.java#L329). Altogether it implies that the `ClusterConfigurationLoader` doesn't return `null` and throws an exception instead, skipping the retry process and failing fast. This scenario is covered by test method `requestConfigurationFromLocatorsShouldThrowExceptionWhenClusterConfigurationServiceIsNotEnabled` within class `ClusterConfigurationLoaderIntegrationTest`, by the way.
Cheers.

jinmeiliao replied:
oh, I see. I thought the exception was wrapped in a CliFunctionResult.

jujoramos replied:
Thanks @jinmeiliao!! 👍

jinmeiliao commented:
@jujoramos the precheckin has some failures which I am not sure if it's environment related. Just to be safe, can you rebase your changes to the current develop and then do a force commit to trigger the precheckin to run again with all the latest fixes? Thanks!

jujoramos commented:
Done @jinmeiliao, I've rebased against develop and executed a `push --force`.
Hope everything's fine now, please let me know if something's missing!.
Cheers.

jujoramos commented:
Hello @jinmeiliao 

I found where the problem is...
In the previous version of the `GetClusterConfigurationFunction` we were catching and logging only instances of the `IOException` class when calling the method `clusterConfigurationService.createConfigurationResponse(Set<String> groups)`. That particular method, however, never throws that `Exception` anyway, so I directly removed the `throws` clause from the declaration and modified the `GetClusterConfigurationFunction` as below:

```
try {
  ConfigurationResponse response =
    clusterConfigurationService.createConfigurationResponse(groups);
  context.getResultSender().lastResult(response);
} catch (Exception exception) {
  logger.warn("Unable to retrieve the cluster configuration", exception);
  context.getResultSender().lastResult(exception);
}
```
​
This means that we are catching and logging (using `log-level=ERROR`) every single exception that might occur while creating the `ConfigurationResponse`. That said, the class `MemberFunctionStreamingMessage` already catches all exceptions thrown by the function execution, but it logs them using `log-level=DEBUG`:

```
...
  try {
...
      this.functionObject.execute(context);
      if (!this.replyLastMsg && this.functionObject.hasResult()) {
        throw new FunctionException(
            LocalizedStrings.ExecuteFunction_THE_FUNCTION_0_DID_NOT_SENT_LAST_RESULT
                .toString(functionObject.getId()));
      }
      stats.endFunctionExecution(start, this.functionObject.hasResult());
...
    } catch (Exception exception) {
      if (logger.isDebugEnabled()) {
        logger.debug("Exception occurred on remote member while executing Function: {}",
            this.functionObject.getId(), exception);
      }
      stats.endFunctionExecutionWithException(this.functionObject.hasResult());
      rex = new ReplyException(exception);
      replyWithException(dm, rex);
      // thr = e.getCause();
...
```

This is the reason why the test wasn't failing before: the `LockServiceDestroyedException` was being thrown way before my changes were applied, but it was logged using `log-level=DEBUG` (it seems to be expected at this point of the execution), because of that the `LogConsumer` class wasn't even receiving the event and, thus, no suspicious strings were logged at all so the test didn't fail. Considering that the exception is expected at this point, I've changed `logger.error()` to `logger.warn()` within the function implementation and everything works fine.
The changes have been rebased and committed already, please let me know if you want me to change something else (didn't squash everything into one single commit to keep the history, but I can do it if you want).
Cheers.
